### PR TITLE
fix(table): rework data source

### PR DESCRIPTION
**Pull request by pweyrich**
#### Must be done by Terra

Testing

-   [ ] Person 1 (mandatory)
-   [ ] Person 2 (mandatory)

Terra Basic Plugin

-   [ ] Adapt changes (optional: relevant for version updates, global design changes, l10n)

---

#### Must be done by every developer

Please inform a member of Terra (@plentymarkets/team-terra) to get the upper part of the checklist done (with urgency or deadline).

Documentation

-   [x] Changelog (optional: relevant for every change which influences the API)
-   [ ] Example (updated or created)
-   [x] JSDoc (optional: relevant for every change which influences the API)

Testing

-   [x] Unit Test (updated or created)
Bildschirmfoto 2020-10-16 um 10 35 38

### CODE REVIEW

**Review by TimWieder — DISMISSED** [review]

File: src/lib/components/table/terra-table-data-source.spec.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        expect(dataSource.disconnect).not.toThrow();
+    });
+
+    it('should be able to manually set data', () => {
+        const newData: Array<any> = [{ foo: 'bar' }];
+        dataSource.data = newData;
+        expect(dataSource.data).toEqual(newData);
+    });
+
+    it('should notify the table if data has been updated manually', () => {
```
Comment (pweyrich) [inline src/lib/components/table/terra-table-data-source.spec.ts:38]:
```suggestion
    it('should notify the table when data has been updated manually', () => {
```

File: src/lib/components/table/terra-table-data-source.spec.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        expect(data).toEqual([]);
+        dataSource.connect().subscribe((d: Array<any>) => (data = d));
+
+        const newData: Array<any> = [{ foo: 'bar' }];
+        dataSource.data = newData;
+
+        expect(data).toEqual(newData);
+    });
+
+    it('should start a request if search is called', () => {
```
Comment (pweyrich) [inline src/lib/components/table/terra-table-data-source.spec.ts:49]:
```suggestion
    it('should start a request when search is called', () => {
```

File: src/lib/components/table/terra-table-data-source.spec.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        beforeEach(() => {
+            filter = new TerraFilter();
+            dataSource.filter = filter;
+        });
+
+        it('should be able to assign a filter', () => {
+            expect(dataSource.filter).toBe(filter);
+        });
+
+        it('should start a request if a search is triggered via the filter', () => {
```
Comment (pweyrich) [inline src/lib/components/table/terra-table-data-source.spec.ts:69]:
```suggestion
        it('should start a request when a search is triggered via the filter', () => {
```

File: src/lib/components/table/terra-table-data-source.spec.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                sortChange: new EventEmitter()
+            } as MatSort;
+            dataSource.sort = sort;
+        });
+
+        it('should be able to assign a MatSort instance', () => {
+            expect(dataSource.sort).toBe(sort);
+        });
+
+        it('should start a request if MatSort is given, data is present and a sort event occurs', fakeAsync(() => {
```
Comment (pweyrich) [inline src/lib/components/table/terra-table-data-source.spec.ts:103]:
```suggestion
        it('should start a request when MatSort is given, data is present and a sort event occurs', fakeAsync(() => {
```

File: src/lib/components/table/terra-table-data-source.spec.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                page: new EventEmitter()
+            } as MatPaginator;
+            dataSource.paginator = paginator;
+        });
+
+        it('should be able to assign a paginator instance', () => {
+            expect(dataSource.paginator).toBe(paginator);
+        });
+
+        it('should start a request if a paginator is available, data is present and the page or pageSize has changed', fakeAsync(() => {
```
Comment (pweyrich) [inline src/lib/components/table/terra-table-data-source.spec.ts:135]:
```suggestion
        it('should start a request when a paginator is available, data is present and the page or pageSize has changed', fakeAsync(() => {
```

**Review by TimWieder — DISMISSED** [review]

**Review by MFrank84 — APPROVED** [review]
Tested 👍

**Review by MFrank84 — APPROVED** [review]

**Review by TimWieder — APPROVED** [review]

**Review by Neill83 — APPROVED** [review]

**Pull request merged**